### PR TITLE
Add ScriptLoadBalancer

### PR DESCRIPTION
**Pull request by tgianos**
This PR includes a new implementation of the `ClusterLoadBalancer` interface which allows system administrators to provide a location for a script which can be downloaded and evaluated within the JVM for selecting a cluster. Currently it supports `JavaScript` and `Groovy` as script languages. More can be added by adding jars to the classpath.

The documentation has been updated with more details.

### CODE REVIEW

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java [hunk 1]
```diff
@@ -331,10 +342,81 @@ private void setRuntimeEnvironment(
...
+                        } else {
+                            this.registry.counter(
+                                this.loadBalancerId
+                                    .withTag(
+                                        LOAD_BALANCER_CLASS_TAG,
+                                        loadBalancerClass
+                                    )
+                                    .withTag(
+                                        LOAD_BALANCER_STATUS_TAG,
+                                        LOAD_BALANCER_STATUS_FAILURE
```
Comment (mprimi) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java]: Here returning null is considered FAILURE in terms of metrics, but maybe something like "NO_PREFERENCE" is more accurate?

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/RandomizedClusterLoadBalancerImpl.java [hunk 2]
```diff
@@ -52,4 +54,12 @@ public Cluster selectCluster(final List<Cluster> clusters) throws GenieException
         final Random rand = new Random();
         return clusters.get(Math.abs(rand.nextInt(clusters.size())));
     }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getOrder() {
+        return Ordered.LOWEST_PRECEDENCE;
```
Comment (mprimi) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/RandomizedClusterLoadBalancerImpl.java]: We want this to always go last. What happens if another LB implementation also uses LOWEST_PRECEDENCE?
PriorityOrdered only works the other way around (giving higher precedence if two objects have the same precedence value).

File: genie-core/src/test/java/com/netflix/genie/core/configs/ServicesConfigTest.java [hunk 3]
```diff
@@ -263,8 +263,7 @@ public JobSubmitterService jobSubmitterService(
         final ApplicationEventPublisher eventPublisher,
         final ApplicationEventMulticaster eventMulticaster,
         final List<WorkflowTask> workflowTasks,
-        @Qualifier("jobsDir")
-        final Resource genieWorkingDir,
+        @Qualifier("jobsDir") final Resource genieWorkingDir,
```
Comment (mprimi) [inline genie-core/src/test/java/com/netflix/genie/core/configs/ServicesConfigTest.java]: Ok this is _my_ personal OCD, but this this line change (and a couple others) is unrelated to this feature/patch.
Sneaking it in can make it harder to track down changes and their reason via `git blame`.
Beside, isn't it generally recommended to have each annotation on a new line?
e.g.: 
>[each annotation is listed on a line of its own (that is, one annotation per line)](https://google.github.io/styleguide/javaguide.html#s4.8.5-annotations)

> [When used on a declaration, each annotation often appears, by convention, on its own line](https://docs.oracle.com/javase/tutorial/java/annotations/basics.htm)

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java [hunk 4]
```diff
@@ -331,10 +342,81 @@ private void setRuntimeEnvironment(
...
         try {
-            log.info("Selecting cluster for job {}", jobRequest.getId());
-            final Cluster cluster
-                = this.clusterLoadBalancer.selectCluster(this.clusterService.chooseClusterForJobRequest(jobRequest));
-            log.info("Selected cluster {} for job {}", cluster.getId(), jobRequest.getId());
+            log.info("Selecting cluster for job {}", jobRequest.getId().orElse(NO_ID_FOUND));
+            final List<Cluster> clusters = this.clusterService.chooseClusterForJobRequest(jobRequest);
+            Cluster cluster = null;
+            if (clusters.isEmpty()) {
+                throw new GeniePreconditionException(
```
Comment (mprimi) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java]: How about a metric/tag for this outcome?

File: genie-docs/src/docs/asciidoc/features/_clusterLoadBalancing.adoc [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+===== Script Contract
+
+The contract between the script and the Java code is as follows:
+
+.Script Inputs
+|===
+|Parameter |Description
+
+|clusters
+|JSON array of cluster objects (serialized as a string) of the clusters to be evaluated
```
Comment (mprimi) [inline genie-docs/src/docs/asciidoc/features/_clusterLoadBalancing.adoc]: As of now, this is guaranteed to be non-empty. Should we make it explicit here and in Javadoc?

File: genie-web/src/main/java/com/netflix/genie/GenieWeb.java [hunk 6]
```diff
@@ -41,8 +41,8 @@
 @EnableAspectJAutoProxy
 public class GenieWeb {
 
-    protected static final String SPRING_CONFIG_LOCATION = "spring.config.location";
-    protected static final String USER_HOME_GENIE = "${user.home}/.genie/";
+    static final String SPRING_CONFIG_LOCATION = "spring.config.location";
+    static final String USER_HOME_GENIE = "${user.home}/.genie/";
```
Comment (mprimi) [inline genie-web/src/main/java/com/netflix/genie/GenieWeb.java]: Changes in this file are also not strictly related to this commit.
Splitting them into a different commit (and then not squashing on merge) would make the history of these lines more clear.

File: genie-web/src/main/java/com/netflix/genie/web/configs/ServicesConfig.java [hunk 7]
```diff
@@ -235,8 +235,7 @@ public JobKillService jobKillService(
         final Executor executor,
         final JobsProperties jobsProperties,
         final ApplicationEventPublisher eventPublisher,
-        @Qualifier("jobsDir")
-        final Resource genieWorkingDir,
+        @Qualifier("jobsDir") final Resource genieWorkingDir,
```
Comment (mprimi) [inline genie-web/src/main/java/com/netflix/genie/web/configs/ServicesConfig.java]: Same as above for style/history.

File: genie-web/src/main/java/com/netflix/genie/web/configs/ServicesConfig.java [hunk 8]
```diff
@@ -286,10 +285,8 @@ public GenieFileTransferService genieFileTransferService(
     @Bean
     public GenieFileTransferService cacheGenieFileTransferService(
         final FileTransferFactory fileTransferFactory,
-        @Value("${genie.file.cache.location}")
```
Comment (mprimi) [inline genie-web/src/main/java/com/netflix/genie/web/configs/ServicesConfig.java]: Same as above for style/history.

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java [hunk 9]
```diff
@@ -331,10 +342,81 @@ private void setRuntimeEnvironment(
     private Cluster getCluster(final JobRequest jobRequest) throws GenieException {
         final long start = System.nanoTime();
         try {
-            log.info("Selecting cluster for job {}", jobRequest.getId());
-            final Cluster cluster
-                = this.clusterLoadBalancer.selectCluster(this.clusterService.chooseClusterForJobRequest(jobRequest));
-            log.info("Selected cluster {} for job {}", cluster.getId(), jobRequest.getId());
+            log.info("Selecting cluster for job {}", jobRequest.getId().orElse(NO_ID_FOUND));
+            final List<Cluster> clusters = this.clusterService.chooseClusterForJobRequest(jobRequest);
```
Comment (mprimi) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java]: Should this list be immutable to avoid plugins messing with it (as of now, I think LBs can add and remove elements).

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java [hunk 10]
```diff
@@ -331,10 +342,81 @@ private void setRuntimeEnvironment(
...
+                                    .withTag(
+                                        LOAD_BALANCER_CLASS_TAG,
+                                        loadBalancerClass
+                                    )
+                                    .withTag(
+                                        LOAD_BALANCER_STATUS_TAG,
+                                        LOAD_BALANCER_STATUS_SUCCESS
+                                    )
+                            ).increment();
+                            cluster = selectedCluster;
```
Comment (mprimi) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java]: Should we check that cluster is in the original list of clusters?

File: genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+    static final String STATUS_TAG_NOT_CONFIGURED = "not configured";
+    static final String STATUS_TAG_FOUND = "found";
+    static final String STATUS_TAG_FAILED = "failed";
+    static final String EXCEPTION_TAG_KEY = "exception";
+
+    private static final long DEFAULT_TIMEOUT_LENGTH = 5_000L;
+    private static final Charset UTF_8 = Charset.forName("UTF-8");
+    private static final String SLASH = "/";
+    private static final String PERIOD = ".";
+    private static final int ORDER = Ordered.LOWEST_PRECEDENCE - 1;
```
Comment (mprimi) [inline genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java]: Should this be a property?
If someone wants to use this, they have no chance to layer something in between Script and Random

File: genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    private final AtomicBoolean isConfigured = new AtomicBoolean();
+    private final ScriptEngineManager scriptEngineManager = new ScriptEngineManager();
+
+    private final AsyncTaskExecutor taskExecutor;
+    private final GenieFileTransferService fileTransferService;
+    private final Environment environment;
+    private final ObjectMapper mapper;
+    private final Registry registry;
+
+    private CompiledScript script;
```
Comment (mprimi) [inline genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java]: Should this be an AtomicReference?

File: genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+     */
+    @Override
+    public int getOrder() {
+        return ORDER;
+    }
+
+    /**
+     * Check if the script file needs to be refreshed.
+     */
+    public void refresh() {
```
Comment (mprimi) [inline genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java]: Maybe just paranoia, but would it be worth wrapping the entire refresh() body into a catch all?
Just to make sure we don't end up killing the executor thread.
(I think the current code is already not supposed to throw, but it wouldn't hurt to add an extra layer)

File: genie-web/src/test/resources/application-ci.yml [hunk 14]
```diff
@@ -30,6 +30,10 @@ genie:
   health:
     maxCpuLoadPercent: 100
   jobs:
+    clusters:
```
Comment (mprimi) [inline genie-web/src/test/resources/application-ci.yml]: `application.yml` above does not rely on default property values, whereas this and the next conf do. Is this intentional.

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java (hunk 9 above)
Comment (tgianos) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java]: Yeah seems a better defensive way to do this. Will change

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java (hunk 4 above)
Comment (tgianos) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java]: Sure.

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java (hunk 10 above)
Comment (tgianos) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java]: Sure. And added counter.

File: genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java (hunk 1 above)
Comment (tgianos) [inline genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java]: agreed

File: genie-docs/src/docs/asciidoc/features/_clusterLoadBalancing.adoc (hunk 5 above)
Comment (tgianos) [inline genie-docs/src/docs/asciidoc/features/_clusterLoadBalancing.adoc]: done

File: genie-web/src/main/java/com/netflix/genie/GenieWeb.java (hunk 6 above)
Comment (tgianos) [inline genie-web/src/main/java/com/netflix/genie/GenieWeb.java]: Will try to keep this in mind for the future. For now this is already squashed so too late

File: genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java (hunk 11 above)
Comment (tgianos) [inline genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java]: I agree with this. Lets do it as another feature in a different PR rather than bundling it in this one in the interest of getting this out.

File: genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java (hunk 12 above)
Comment (tgianos) [inline genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java]: Done

File: genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java (hunk 13 above)
Comment (tgianos) [inline genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java]: I moved the try up as high as possible while still giving the finally block access to the start time and map of strings. The try block catches every checked exception and `RuntimeException` so nothing can get out of the block.

File: genie-web/src/test/resources/application-ci.yml (hunk 14 above)
Comment (tgianos) [inline genie-web/src/test/resources/application-ci.yml]: It is cause I don't want someone starting up Genie without knowledge of the `ScriptLoadBalancer` and seeing a bunch of Stack Traces periodically cause the system couldn't find the script. But for the integration test I wanted to make sure that the system could still run even if the `ScriptLoadBalancer` was enabled and not configured correctly. I could take it back out but it doesn't really hurt anything for the testing. Ideally this would pull a script to actually test running but the UnitTest already does that.

File: genie-web/src/test/groovy/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancerSpec.groovy [hunk 15]
```diff
@@ -254,5 +254,6 @@ class ScriptLoadBalancerSpec extends Specification {
         type         | file
         "JavaScript" | this.class.getResource("loadBalance.js").file
         "Groovy"     | this.class.getResource("loadBalance.groovy").file
+//        "Python"     | this.class.getResource("loadBalance.py").file
```
Comment (mprimi) [inline genie-web/src/test/groovy/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancerSpec.groovy]: Maybe forgot to delete this line?

File: build.gradle [hunk 16]
```diff
@@ -94,7 +94,13 @@ configure(javaProjects) {
             dependency("org.apache.commons:commons-exec:1.3")
             dependency("org.bitbucket.b_c:jose4j:0.5.4")
             dependency("org.dbunit:dbunit:2.5.3")
+            dependencySet(group: "org.jruby", version:"9.1.12.0") {
```
Comment (mprimi) [inline build.gradle]: Are JRuby and Jython really necessary?

Comment (tgianos) [inline build.gradle]: This is the dependency management section. They're not being added to the project in any way right now other than the versions are locked if we decide to ever add them. In this case I was testing the `ScriptEngine` implementations they had and decided to leave the dependency lock in just in case.
